feat(segment_regquery): segment to query and display a registry key on Windows.

memcpy-rand-rand-rand opened the pull request:
### Prerequisites

- [X] I have read and understood the `CONTRIBUTING` guide
- [X] The commit message follows the [conventional commits][cc] guidelines
- [X] Tests for the changes have been added (for bug fixes/features)
- [X] Docs have been added / updated (for bug fixes/features)

### Description

New segment, tests and docs which will allow the user to specify a Windows registry path and key and display the result (as the string returned if registry key type is REG_SZ or as an 0x%08X capital hex number for type REG_DWORD).  Ability to specify different behaviours for failure conditions (do not display, use supplied fallback string or show debug info).

build: tags used by OS to supply bypass implementation for non-Windows.

Also, in this PR - minor fix to the segment contribution doc

Hadn't used CC msgs while playing about in the fork, most recent commits updated to do so.

<!---

Tips:

If you're not comfortable with working with Git, we're working a guide (https://ohmyposh.dev/docs/contributing_git) to help you out.
Oh My Posh advises GitKraken (https://www.gitkraken.com/invite/nQmDPR9D) as your preferred cross platform Git GUI power tool.

-->

[cc]: https://www.conventionalcommits.org/en/v1.0.0/#summary

lnu committed:
feat(sysinfo): sysinfo segment

cpu, mem, load, ...

memcpy-rand-rand-rand committed:
First add of registry query segment.

memcpy-rand-rand-rand committed:
Add regquery to schema enum

memcpy-rand-rand-rand committed:
feat(regquery): initial creation

feat(regquery): initial reg query code

feat(regquery): reading values out of the registry for real

feat(regquery): root key and fallback

docs: regquery first pass markdown doc page

docs: regquery doc page tweaks

feat(regquery): doc updates, non-windows platform handling

JanDeDobbeleer committed:
fix: use reference colors in transient prompt

JanDeDobbeleer committed:
theme: replace \ufa1e with \ufbab

closes #1239

JanDeDobbeleer committed:
chore: show version and log on debug

JanDeDobbeleer committed:
chore: ask debug log for bug report

JanDeDobbeleer committed:
refactor: provide log output directly on debug

memcpy-rand-rand-rand committed:
first pass to get reg key open/query handling going.
- open key
- query key for size and type

JanDeDobbeleer committed:
refactor: remove trace struct

JanDeDobbeleer committed:
chore: show total duration in debug

shiipou committed:
chore: fix not unique id

memcpy-rand-rand-rand committed:
test flow reading values out of the registry for real.

JanDeDobbeleer committed:
chore: fix bug issue form

JanDeDobbeleer committed:
docs(az): align sample with template functionality

JanDeDobbeleer committed:
feat(cmd): add tooltip functionality

memcpy-rand-rand-rand committed:
add root key choice for config
add fallback routes for errors getting the value of the key

JanDeDobbeleer committed:
feat(cmd): transient prompt

memcpy-rand-rand-rand committed:
first pass markdown doc page

memcpy-rand-rand-rand committed:
Doc page tweaks

memcpy-rand-rand-rand committed:
add "non-windows" segment code to short-circuit and never show the segment
docs tweak (fix backtick quotes, change segment bg colour)

memcpy-rand-rand-rand committed:
Merge branch 'JanDeDobbeleer:main' into main

memcpy-rand-rand-rand committed:
feat(regquery): rename file - replace dash with underscore

memcpy-rand-rand-rand committed:
rename file - replace dash with underscore

TedCrocker committed:
feat(git): add symbol to the branch name when truncating

allcontributors[bot] committed:
docs: update docs/docs/contributors.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

JanDeDobbeleer committed:
fix(cmd): allow all types of path separators

memcpy-rand-rand-rand committed:
Update contributing-segment.md

fix enabled() function to be "return true"

memcpy-rand-rand-rand committed:
docs: Update contributing-segment.md, return true from enabled

billyrayxbox committed:
feat(regquery): refactor registry query code

billyrayxbox committed:
refactor registry query code - moved out of enabled() and getting out of my Turbo Pascal zone to make it a short-circuit rather than nested if/else structure!
remove root key requirement - parse this from the supplied string, update doc page and schema

billyrayxbox committed:
Merge branch 'main' of https://github.com/memcpy-rand-rand-rand/oh-my-posh

memcpy-rand-rand-rand committed:
add test code - primarily exercises registry key get logic

memcpy-rand-rand-rand committed:
feat(regquery): registry query logic tests

feat(regquery): fix non-windows platform files to support build and tests

feat(regquery): fix incorrect version checkin of file

feat(regquery): change test function names for clarity in test output

lnu committed:
docs: sysinfo not ordered correctly in sidebar

memcpy-rand-rand-rand committed:
fix non-windows platform files to support build and tests

memcpy-rand-rand-rand committed:
fix incorrect version checkin of file

memcpy-rand-rand-rand committed:
Merge branch 'JanDeDobbeleer:main' into main

memcpy-rand-rand-rand committed:
change test function names for clarity in test output

memcpy-rand-rand-rand committed:
Merge branch 'main' of https://github.com/memcpy-rand-rand-rand/oh-my-posh

memcpy-rand-rand-rand committed:
feat(regquery): remove dead registry root property reference

memcpy-rand-rand-rand committed:
remove dead registry root property reference

memcpy-rand-rand-rand committed:
doc: modify reg query example

memcpy-rand-rand-rand committed:
doc: modify reg query example

memcpy-rand-rand-rand committed:
change example regquery for docs

memcpy-rand-rand-rand committed:
feat(regquery): windows registry query segment

now using conventional commit

memcpy-rand-rand-rand committed:
feat(regquery): windows registry query segment

now using conventional commit

memcpy-rand-rand-rand committed:
Merge branch 'main' of https://github.com/memcpy-rand-rand-rand/oh-my-posh

memcpy-rand-rand-rand committed:
docs: add return to enabled() in contrib segment doc

memcpy-rand-rand-rand committed:
docs: add return to enabled() in contrib segment doc

memcpy-rand-rand-rand commented:
Any idea how I can fix the commitlint issues in retrospect?

JanDeDobbeleer commented:
@memcpy-rand-rand-rand if you use `git rebase main -i` (ensure main is up to date with ours), you can squash all commits and reword the message.

memcpy-rand-rand-rand committed:
docs: update return true

memcpy-rand-rand-rand committed:
feat(regquery): merge fix

memcpy-rand-rand-rand commented:
there's a reasonable chance I'm going to tie things in knots due to my weak git-fu, looking at a cleaner way to get these changes committed with appropriate comments.

JanDeDobbeleer commented:
I can do that for you my friend. I'll try to push a branch later this evening when I review.

memcpy-rand-rand-rand commented:
many thanks Jan - I'll reopen it, don't want you spending too many cycles on my lack of git skillz tho :)

JanDeDobbeleer commented:
@memcpy-rand-rand-rand don't worry. I've got this 😉

memcpy-rand-rand-rand commented:
💓

memcpy-rand-rand-rand committed:
fix: markdown lint issues in segmentregquery doc

JanDeDobbeleer commented on src/segment.go:
```diff
@@ -130,6 +130,8 @@ const (
 	// PHP writes which php version is currently active
 	PHP SegmentType = "php"
+	// RegQuery queries the Windows registry.
+	RegQuery SegmentType = "regquery"
```
I would rename this to `winreg`, it's about the registry, the fact that we query it is an implementation detail.

JanDeDobbeleer commented on src/segment_regquery.go:
```diff
@@ -0,0 +1,29 @@
+
+	// Properties defining fallback behaviour in case the query fails.
+	QueryFailBehaviour      Property = "query_fail_behaviour"       // what to do, can be one of "hide_segment", "display_fallback_string", "show_debug_info"
+	QueryFailFallbackString Property = "query_fail_fallback_string" // what to display if above is "display_fallback_string"
```
I would drop this and hide the segment on error or no value. If we move the actual implementation code to `environment_windows.go`, we can use oh-my-posh's native debug features to let people know what happens in case the segment isn't shown.

memcpy-rand-rand-rand replied:
the fallback string method is useful - in a particular case for which I was writing this, the lack of existence the key in the registry defines a default state that it's useful to display.

totally agree on the debug though, much better to have that logged via the env's debug.

JanDeDobbeleer replied:
I agree, but the I'd log that straight in the debug log. But I think that's also what you meant.

memcpy-rand-rand-rand replied:
Actually, the fallback string is intended for the segment, not debug.  Use case is that I am querying the registry for what a certain toolchain puts in there based on how I configure it.  When there is a value there, it defines the debug environment I am in, when it's not there it will be using the live environment.  As a cmd prompt user, I want to be able to clearly see both scenarios (non-existence of the segment is not explicit enough).

There are certainly other cases where the segment could just not show.  I'd be happy to share some more details on specifics privately.

memcpy-rand-rand-rand replied:
Aha - looks like this could be cleaner done via template system, grokking how that works now.

memcpy-rand-rand-rand replied:
done via templates

JanDeDobbeleer commented on src/segment_regquery.go:
```diff
@@ -0,0 +1,29 @@
+)
+
+func (r *regquery) string() string {
+	newText := r.content
```
this needs to make use of templates (see the AZ segment for details for example)

memcpy-rand-rand-rand replied:
done

JanDeDobbeleer commented on src/segment_regquery_others.go:
```diff
@@ -0,0 +1,7 @@
+package main
+
+func (r *regquery) enabled() bool {
+	return false
```
Rather than split this, I would a simply do the following in this case:

```golang
if r.env.getRuntimeGOOS() != windowsPlatform {
    return false
}
```

We will never implement the registry segment for any other platform.

memcpy-rand-rand-rand replied:
done

JanDeDobbeleer commented on src/segment_regquery_others_test.go:
```diff
@@ -0,0 +1,14 @@
+	"github.com/alecthomas/assert"
+)
+
+func TestRegQueryNonWindows(t *testing.T) {
```
This file can then be removed.

memcpy-rand-rand-rand replied:
done with pleasure :)

JanDeDobbeleer commented on src/segment_regquery_windows.go:
```diff
@@ -0,0 +1,143 @@
+	"strings"
+	"unsafe"
+
+	"golang.org/x/sys/windows"
```
I only want to see code like this in `environment_windows.go`, it's not testable (at least not without Windows).

memcpy-rand-rand-rand replied:
done

JanDeDobbeleer commented on src/segment_regquery_windows.go:
```diff
@@ -0,0 +1,143 @@
+//   true and the retrieved value formatted into a string if successful.
+//   false and the string will be the error
+//
+func (r *regquery) getRegistryKeyValue(regPath string, regKey string) (bool, string) {
```
this function and all it's required logic needs to move to `environment_windows.go`

memcpy-rand-rand-rand replied:
done

JanDeDobbeleer commented on src/segment_regquery_windows.go:
```diff
@@ -0,0 +1,143 @@
+
+	regRootHKeyHandle := r.getHKEYValueFromAbbrString(regPathParts[0])
+	if regRootHKeyHandle == 0 {
+		return false, fmt.Sprintf("Error, Supplied root HKEY value not valid: '%s'", regPathParts[0])
```
in `environment_windows.go`, you can make use of `env.log(Error, "getRegistryKeyValue", err.Error())`

memcpy-rand-rand-rand replied:
splendid - logging and trace code added.

JanDeDobbeleer commented on src/segment_regquery_windows_test.go:
```diff
@@ -0,0 +1,94 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func TestRegQueryGetRegistryFunction(t *testing.T) {
```
this is an integration test, not a functionality test for the segmenmt. It's good, but needs to move to the environment tests (on windows)

memcpy-rand-rand-rand replied:
test code still needs sorting out - currently commented out

JanDeDobbeleer commented:
Do not worry about commitlint, I can squash here and we can always merge it of all other tests succeed :-)

I added quite a few architectural requests, don't get discouraged as it's a great addition! If you need any help, reach out!

memcpy-rand-rand-rand commented:
thanks a lot for the review, love the idea of abstracting the platform stuff into env.  Will crack into that.

memcpy-rand-rand-rand committed:
refactor: first pass move registry query code into environmemnt

change name of func
move away from bool return to existing error methodology
treat error and registy value as separate return values
add trace
add debug logging
add mocked version
add "not implemented" version

memcpy-rand-rand-rand committed:
refactor: class rename regquery to winreg

code changes only

memcpy-rand-rand-rand committed:
refactor: regquery to winreg code

memcpy-rand-rand-rand committed:
refactor: regquery to winreg name, schema and docs

memcpy-rand-rand-rand committed:
fix: remove build directive

memcpy-rand-rand-rand committed:
Merge branch 'JanDeDobbeleer:main' into winreg-refactor

memcpy-rand-rand-rand committed:
refactor(winreg): short-circuit enabled() to false on non-windows

memcpy-rand-rand-rand committed:
fix(winreg): debug output function name

memcpy-rand-rand-rand committed:
fix(winreg): md lint line lengths

memcpy-rand-rand-rand committed:
refactor(winreg): use templates to handle fallback behaviour

memcpy-rand-rand-rand committed:
Merge branch 'winreg-refactor' of https://github.com/memcpy-rand-rand-rand/oh-my-posh into winreg-refactor

memcpy-rand-rand-rand committed:
Merge branch 'JanDeDobbeleer:main' into winreg-refactor

memcpy-rand-rand-rand committed:
Merge branch 'JanDeDobbeleer:main' into main

memcpy-rand-rand-rand committed:
Merge pull request #1 from memcpy-rand-rand-rand/winreg-refactor

refactor(winreg): based on feedback from Jan

memcpy-rand-rand-rand commented:
winreg-refactor branch pulled into main with suggested PR changes.
refactor of test code pending

JanDeDobbeleer commented:
I have no pointers, looking forward to the tests. Well done.

JanDeDobbeleer commented on src/environment_windows.go:
```diff
@@ -99,3 +103,93 @@ func (env *environment) getCachePath() string {
+//   false and the string will be the error
+//
+func (env *environment) getWindowsRegistryKeyValue(regPath string, regKey string) (string, error) {
+	env.trace(time.Now(), "getWindowsRegistryKeyValue", regPath, regKey)
```
love that you also added this

billyrayxbox replied:
love me some perf metrics :)

billyrayxbox replied:
(same guy, different account, forgot I'm on the work laptop!)

billyrayxbox committed:
fix(winreg): lint errors

memcpy-rand-rand-rand committed:
Merge branch 'JanDeDobbeleer:main' into main

memcpy-rand-rand-rand commented:
fixed code lint errors, not sure how to fix the commitlint wording error the previous run highlighted, @JanDeDobbeleer

JanDeDobbeleer commented:
> not sure how to fix the commitlint wording error

@memcpy-rand-rand-rand don't worry about that. I'm squashing this one anyways ;-)

memcpy-rand-rand-rand commented:
and of course I missed a couple of sites of the code lint error - fix going in shortly.

billyrayxbox committed:
fix(winreg): lint errors in interface and unix impl

billyrayxbox committed:
Merge branch 'main' of https://github.com/memcpy-rand-rand-rand/oh-my-posh

memcpy-rand-rand-rand commented:
> and of course I missed a couple of sites of the code lint error - fix going in shortly.

fixed, @JanDeDobbeleer

memcpy-rand-rand-rand commented:
Fyi, I'm thinking about a refactor of the registry key function in environment, so it's more useful to others, it may end up in this PR, depending on my bandwidth!

JanDeDobbeleer commented:
@memcpy-rand-rand-rand I had to create a new branch due to quite a bit of conflicts, but it's in. So you can do that including adding some integration tests for that.

JanDeDobbeleer commented:
@All-contributors please add @memcpy-rand-rand-rand for code and doc

memcpy-rand-rand-rand commented:
whoop!  thanks!
